File: 프로그래머스/레벨2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.java
import java.util.*;

class Solution {
    private String noteConverter(String melody) {
        StringBuilder sb = new StringBuilder();
        for (char note : melody.toCharArray()) {
            if (note == '#') {
                int lastIdx = sb.length() - 1;
                char newNote = Character.toLowerCase(sb.charAt(lastIdx));
                sb.setCharAt(lastIdx, newNote);                
            } else {
                sb.append(note);
            }
        }
        return sb.toString();
    }

    private int minuteConverter(String timeString) {
        String[] parts = timeString.split(":");
        int hour = Integer.parseInt(parts[0]);
        int minute = Integer.parseInt(parts[1]);
        return 60 * hour + minute;
    }

    private String melodyRepeater(String melody, int playTime) {
        return melody.repeat(playTime / melody.length() + 1).substring(0, playTime);
    }    

    public String solution(String m, String[] musicinfos) {
        String convM = noteConverter(m);
        String resultTitle = "(None)";
        int resultTime = 0;

        for (String musicinfo : musicinfos) {
            String[] parts = musicinfo.split(",");
            String start = parts[0];
            String end = parts[1];
            String title = parts[2];
            String melody = parts[3];

            int playTime = minuteConverter(end) - minuteConverter(start);
            String convMelody = noteConverter(melody);
            String repeatedMelody = melodyRepeater(convMelody, playTime);

            if (repeatedMelody.contains(convM) && playTime > resultTime) {
                resultTitle = title;
                resultTime = playTime;
            }
        }

        return resultTitle;
    }
}

File: 프로그래머스/레벨2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.cpp
#include <bits/stdc++.h>

using namespace std;

string noteConverter(const string& melody) {
    string result;
    for (char note : melody) {
        if (note == '#') {
            result.back() = tolower(result.back());
        } else {
            result.push_back(note);
        }
    }
    return result;
}

int minuteConverter(const string& timeString) {
    size_t colonPos = timeString.find(':');
    int hour = stoi(timeString.substr(0, colonPos));
    int minute = stoi(timeString.substr(colonPos + 1));
    return 60 * hour + minute;
}

string melodyRepeater(const string& melody, int playTime) {
    string result;
    int repeatCount = ceil(static_cast<double>(playTime) / melody.size());
    for (int i = 0; i < repeatCount; ++i) {
        result += melody;
    }
    return result.substr(0, playTime);
}

string solution(string m, vector<string> musicinfos) {
    string convM = noteConverter(m);
    string resultTitle = "(None)";
    int resultTime = 0;

    for (const string& musicinfo : musicinfos) {
        stringstream ss(musicinfo);
        string start, end, title, melody;

        getline(ss, start, ',');
        getline(ss, end, ',');
        getline(ss, title, ',');
        getline(ss, melody, ',');

        int playTime = minuteConverter(end) - minuteConverter(start);
        string convMelody = noteConverter(melody);
        string repeatedMelody = melodyRepeater(convMelody, playTime);

        if (repeatedMelody.find(convM) != string::npos && playTime > resultTime) {
            resultTitle = title;
            resultTime = playTime;
        }
    }

    return resultTitle;
}